src/java/org/apache/cassandra/journal/Segments.java

```diff
@@ -68,27 +71,34 @@ Segments<K, V> withCompletedSegment(ActiveSegment<K, V> activeSegment, StaticSeg
         return new Segments<>(newSegments);
     }
 
-    Segments<K, V> withCompactedSegment(StaticSegment<K, V> oldSegment, StaticSegment<K, V> newSegment)
+    Segments<K, V> withCompactedSegments(Collection<StaticSegment<K, V>> oldSegments, Collection<StaticSegment<K, V>> compactedSegments)
     {
-        Invariants.checkArgument(oldSegment.descriptor.timestamp == newSegment.descriptor.timestamp);
-        Invariants.checkArgument(oldSegment.descriptor.generation < newSegment.descriptor.generation);
         Long2ObjectHashMap<Segment<K, V>> newSegments = new Long2ObjectHashMap<>(segments);
-        Segment<K, V> oldValue = newSegments.put(newSegment.descriptor.timestamp, newSegment);
-        Invariants.checkState(oldValue == oldSegment);
+        for (StaticSegment<K, V> oldSegment : oldSegments)
+        {
+            Segment<K, V> oldValue = newSegments.remove(oldSegment.descriptor.timestamp);
+            Invariants.checkState(oldValue == oldSegment);
+        }
+
+        for (StaticSegment<K, V> compactedSegment : compactedSegments)
+        {
+            Segment<K, V> oldValue = newSegments.put(compactedSegment.descriptor.timestamp, compactedSegment);
+            Invariants.checkState(oldValue == null);
+        }
+
         return new Segments<>(newSegments);
     }
 
-    Segments<K, V> withoutInvalidatedSegment(StaticSegment<K, V> staticSegment)
+    boolean contains(long descriptor)
     {
-        Long2ObjectHashMap<Segment<K, V>> newSegments = new Long2ObjectHashMap<>(segments);
-        if (!newSegments.remove(staticSegment.descriptor.timestamp, staticSegment))
-            throw new IllegalStateException();
-        return new Segments<>(newSegments);
+        return segments.containsKey(descriptor);
     }
 
     Iterable<Segment<K, V>> all()
     {
-        return segments.values();
+        List<Segment<K, V>> segments = new ArrayList<>(this.segments.values());
```

Comment:
We should probably not be allocating an entire `ArrayList` and sort it for every call to `all()` with this signature. Doesn't sit well with me. But can be addressed later - I've made an internal TODO for myself.

Reply:
Made a separate method that allocates and sorts.